File: backend/app/utils/mentions.py
"""@mention resolution.

A mention is ``@`` followed by the *display name* of a referenced object. Display
names are not identifiers: tables, semantic models and MCP tools routinely carry
spaces, slashes and dots — "Sales Orders", "Microsoft Fabric / dbo.sales".

A regex therefore cannot delimit a mention on its own. In ``@Sales Orders by
region`` nothing in the string says whether the name is "Sales", "Sales Orders"
or "Sales Orders by region" — that depends entirely on which names exist.
Matching ``[A-Za-z0-9_]+`` just takes the first word and silently truncates every
multi-word name.

So mentions are parsed against a dictionary of known names, longest match wins
(maximum munch — the same rule lexers use). The dictionary is held in a trie, so
a lookup costs O(length of the matched name) rather than O(number of names): an
agent with 5k tables parses at the same speed as one with five.

This mirrors ``frontend/utils/mentions.ts``; the two must stay in step, so the
behavioural cases are pinned by tests/unit/test_mentions.py.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Iterable, List, Optional, Sequence, Tuple

# Identifier-shaped fallback: what a mention looks like with no dictionary.
_BARE_MENTION_RE = re.compile(r"[A-Za-z_][A-Za-z0-9_]*(?:[.\-][A-Za-z0-9_]+)*")

# A name needs quoting when its own characters would not survive a bare parse.
_NEEDS_QUOTES_RE = re.compile(r"[\s\-.\"]")

# A dictionary match must end on a word boundary, so "Sales" does not chip the
# head of "@SalesForce".
_WORD_CHAR_RE = re.compile(r"[A-Za-z0-9_]")


@dataclass
class _TrieNode:
    children: dict = field(default_factory=dict)
    name: Optional[str] = None  # canonical display name; set on terminal nodes


class MentionMatcher:
    """Display names indexed for longest-prefix lookup.

    Matching is case-insensitive; the canonical casing from the dictionary is
    returned, so a resolved mention carries the object's real name regardless of
    how it was written.
    """

    def __init__(self, names: Iterable[Optional[str]]):
        self._root = _TrieNode()
        self._size = 0
        for raw in names:
            name = (raw or "").strip()
            if not name:
                continue
            node = self._root
            for ch in name.lower():
                nxt = node.children.get(ch)
                if nxt is None:
                    nxt = _TrieNode()
                    node.children[ch] = nxt
                node = nxt
            # First writer wins, so a stable dictionary order gives stable casing.
            if node.name is None:
                node.name = name
                self._size += 1

    def __len__(self) -> int:
        return self._size

    def match_at(self, text: str, pos: int) -> Optional[Tuple[str, int]]:
        """Longest dictionary name matching ``text`` at ``pos``.

        Returns ``(canonical_name, chars_consumed)`` or None. Walks the trie one
        character at a time, keeping the last terminal seen — so the longest
        candidate wins and the scan stops as soon as no name can extend further.
        """
        node = self._root
        best: Optional[Tuple[str, int]] = None
        for i in range(pos, len(text)):
            nxt = node.children.get(text[i].lower())
            if nxt is None:
                break
            node = nxt
            if node.name is not None:
                after = text[i + 1] if i + 1 < len(text) else ""
                if not after or not _WORD_CHAR_RE.match(after):
                    best = (node.name, i + 1 - pos)
        return best


EMPTY_MATCHER = MentionMatcher([])


@dataclass
class MentionSegment:
    """A run of the source text: literal when ``label`` is None."""

    text: str
    label: Optional[str] = None


def parse_mention_segments(
    text: str, matcher: Optional[MentionMatcher] = None
) -> List[MentionSegment]:
    """Split ``text`` into literal runs and mentions.

    Resolution order at each ``@``:
      1. a quoted label — ``@"Sales Orders"`` — always wins, it is self-delimiting;
      2. the longest name in ``matcher``;
      3. the identifier-shaped fallback, so undictionaried text still resolves.
    """
    src = text or ""
    out: List[MentionSegment] = []
    buffer: List[str] = []

    def flush() -> None:
        if buffer:
            out.append(MentionSegment("".join(buffer)))
            buffer.clear()

    i = 0
    n = len(src)
    while i < n:
        if src[i] == "@":
            rest = src[i + 1 :]

            # 1. Quoted mention.
            if rest[:1] == '"':
                end = rest.find('"', 1)
                if end != -1:
                    label = rest[1:end]
                    flush()
                    out.append(MentionSegment(f'@"{label}"', label))
                    i += 1 + end + 1
                    continue

            # 2. Longest known display name.
            hit = matcher.match_at(src, i + 1) if matcher is not None else None
            if hit is not None:
                name, length = hit
                flush()
                out.append(MentionSegment("@" + src[i + 1 : i + 1 + length], name))
                i += 1 + length
                continue

            # 3. Identifier-shaped fallback.
            word = _BARE_MENTION_RE.match(rest)
            if word:
                flush()
                out.append(MentionSegment("@" + word.group(0), word.group(0)))
                i += 1 + len(word.group(0))
                continue

        buffer.append(src[i])
        i += 1

    flush()
    return out


def extract_mention_labels(
    text: str, matcher: Optional[MentionMatcher] = None
) -> List[str]:
    """Distinct mention labels in ``text``, in order of first appearance."""
    seen = set()
    out: List[str] = []
    for seg in parse_mention_segments(text, matcher):
        if seg.label is None:
            continue
        key = seg.label.lower()
        if key in seen:
            continue
        seen.add(key)
        out.append(seg.label)
    return out


def mention_needs_quotes(name: Optional[str]) -> bool:
    """True when ``name`` must be written ``@"name"`` to survive a bare parse."""
    return bool(name) and bool(_NEEDS_QUOTES_RE.search(name))


def format_mention(name: str) -> str:
    """Canonical source form for a mention of ``name``."""
    return f'@"{name.replace(chr(34), "")}"' if mention_needs_quotes(name) else f"@{name}"


def resolve_mentions(
    text: str, candidates: Sequence[dict]
) -> List[dict]:
    """Resolve the @mentions in ``text`` against ``candidates``.

    ``candidates`` are dicts as returned by
    ``InstructionService.get_available_references`` — each carrying at least
    ``id``, ``type`` and ``name``. Returns one entry per distinct resolved
    mention, in order of first appearance, shaped for
    ``InstructionReferenceCreate``.

    Unresolvable mentions are dropped: they name something that does not exist,
    so there is nothing to link them to.
    """
    by_name: dict = {}
    for cand in candidates or []:
        name = (cand.get("name") or "").strip()
        if not name:
            continue
        by_name.setdefault(name.lower(), cand)

    matcher = MentionMatcher(c.get("name") for c in (candidates or []))

    out: List[dict] = []
    for label in extract_mention_labels(text, matcher):
        cand = by_name.get(label.lower())
        if not cand:
            continue
        out.append(
            {
                "object_type": cand.get("type"),
                "object_id": str(cand.get("id")),
                "relation_type": "mention",
                "display_text": cand.get("name"),
            }
        )
    return out


def canonicalize_mentions(text: str, candidates: Sequence[dict]) -> str:
    """Rewrite every resolvable mention in ``text`` to its canonical source form.

    ``@sales orders`` becomes ``@"Sales Orders"`` — right casing, and quoted so
    the text stays unambiguous for any reader that lacks the dictionary. Text
    that resolves to nothing is left exactly as written.
    """
    matcher = MentionMatcher(c.get("name") for c in (candidates or []))
    known = {(c.get("name") or "").lower() for c in (candidates or [])}
    parts: List[str] = []
    for seg in parse_mention_segments(text, matcher):
        if seg.label is not None and seg.label.lower() in known:
            parts.append(format_mention(seg.label))
        else:
            parts.append(seg.text)
    return "".join(parts)

File: frontend/utils/mentions.ts
/**
 * @mention parsing.
 *
 * A mention is `@` followed by the *display name* of a referenced object. Display
 * names are not identifiers: tables, semantic models and MCP tools routinely carry
 * spaces, slashes and dots — "Sales Orders", "Microsoft Fabric / dbo.sales".
 *
 * A regex therefore cannot delimit a mention on its own. Given the text
 * `@Sales Orders by region`, nothing in the string says whether the name is
 * "Sales", "Sales Orders" or "Sales Orders by region" — that depends entirely on
 * which names exist. Matching `[A-Za-z0-9_]+` just takes the first word and
 * silently truncates every multi-word name.
 *
 * So mentions are parsed against a dictionary of known names, longest match wins
 * (maximum munch — the same rule lexers use for `>>` vs `>`). The dictionary lives
 * in a trie, so a lookup costs O(length of the matched name) instead of
 * O(number of names): an agent with 5k tables parses at the same speed as one with
 * five. Without a dictionary the parser degrades to the historical
 * identifier-shaped match, so callers that have no name list still work.
 *
 * Quoted mentions (`@"Sales Orders"`) are always honored, dictionary or not —
 * that is the form the mention picker inserts, and it is self-delimiting.
 */

/** Identifier-shaped fallback: what a mention looks like with no dictionary. */
const BARE_MENTION_RE = /^[A-Za-z_][A-Za-z0-9_]*(?:[.\-][A-Za-z0-9_]+)*/

/** A name needs quoting when its own characters would not survive a bare parse. */
const NEEDS_QUOTES_RE = /[\s\-."]/

/** Characters that may not immediately follow a dictionary match — a match must
 *  end on a word boundary, so "Sales" does not chip the head of "@SalesForce". */
const WORD_CHAR_RE = /[A-Za-z0-9_]/

interface TrieNode {
  children: Map<string, TrieNode>
  /** Canonical display name, set on terminal nodes. */
  name?: string
}

export interface MentionMatch {
  /** Canonical name as it appears in the dictionary (original casing). */
  name: string
  /** Characters consumed from the text (may differ in case from `name`). */
  length: number
}

/**
 * A dictionary of mentionable display names, indexed for longest-prefix lookup.
 *
 * Matching is case-insensitive; the canonical casing from the dictionary is
 * returned so a chip renders the object's real name regardless of how it was
 * typed.
 */
export class MentionMatcher {
  private root: TrieNode = { children: new Map() }
  readonly size: number

  constructor(names: Iterable<string | null | undefined>) {
    let size = 0
    for (const raw of names) {
      const name = (raw || '').trim()
      if (!name) continue
      let node = this.root
      const lower = name.toLowerCase()
      for (const ch of lower) {
        let next = node.children.get(ch)
        if (!next) {
          next = { children: new Map() }
          node.children.set(ch, next)
        }
        node = next
      }
      // First writer wins, so a stable dictionary order gives stable casing.
      if (node.name === undefined) {
        node.name = name
        size++
      }
    }
    this.size = size
  }

  get isEmpty(): boolean {
    return this.size === 0
  }

  /**
   * Longest dictionary name matching `text` at `pos`, or null.
   *
   * Walks the trie one character at a time, remembering the last terminal node
   * seen — so the longest candidate wins and the scan stops as soon as no name
   * can extend further. Cost is bounded by the longest name in the dictionary,
   * not by its size.
   */
  matchAt(text: string, pos: number): MentionMatch | null {
    let node = this.root
    let best: MentionMatch | null = null
    for (let i = pos; i < text.length; i++) {
      const next = node.children.get(text[i].toLowerCase())
      if (!next) break
      node = next
      if (node.name !== undefined) {
        const after = text[i + 1]
        // Require a word boundary so a short name cannot chip the head of a
        // longer bare token ("Sales" inside "@SalesForce").
        if (after === undefined || !WORD_CHAR_RE.test(after)) {
          best = { name: node.name, length: i + 1 - pos }
        }
      }
    }
    return best
  }
}

/** Empty dictionary — parses with the identifier fallback only. */
export const EMPTY_MENTION_MATCHER = new MentionMatcher([])

/**
 * Build a matcher from whatever reference-ish objects a caller has on hand.
 * Accepts the shapes returned by `/instructions/available-references` and by an
 * instruction's own `references` array.
 */
export function buildMentionMatcher(
  refs: Array<{ name?: string | null; display_text?: string | null } | null | undefined> | null | undefined,
): MentionMatcher {
  if (!refs?.length) return EMPTY_MENTION_MATCHER
  return new MentionMatcher(refs.map(r => r?.name || r?.display_text || null))
}

export type MentionSegment =
  | { type: 'text'; value: string }
  | { type: 'mention'; label: string; /** consumed source text, without the `@` */ raw: string }

/**
 * Split `text` into literal runs and mentions.
 *
 * Resolution order at each `@`:
 *   1. a quoted label — `@"Sales Orders"` — always wins, it is self-delimiting;
 *   2. the longest name in `matcher`;
 *   3. the identifier-shaped fallback, so undictionaried text still chips.
 */
export function parseMentionSegments(
  text: string,
  matcher: MentionMatcher | null = null,
): MentionSegment[] {
  const src = text || ''
  const out: MentionSegment[] = []
  let buffer = ''
  const flush = () => {
    if (buffer) {
      out.push({ type: 'text', value: buffer })
      buffer = ''
    }
  }

  let i = 0
  while (i < src.length) {
    if (src[i] === '@') {
      const rest = src.slice(i + 1)

      // 1. Quoted mention.
      if (rest[0] === '"') {
        const end = rest.indexOf('"', 1)
        if (end !== -1) {
          const label = rest.slice(1, end)
          flush()
          out.push({ type: 'mention', label, raw: `"${label}"` })
          i += 1 + end + 1
          continue
        }
      }

      // 2. Longest known display name.
      const match = matcher?.matchAt(src, i + 1)
      if (match) {
        flush()
        out.push({ type: 'mention', label: match.name, raw: src.substr(i + 1, match.length) })
        i += 1 + match.length
        continue
      }

      // 3. Identifier-shaped fallback.
      const word = rest.match(BARE_MENTION_RE)
      if (word) {
        flush()
        out.push({ type: 'mention', label: word[0], raw: word[0] })
        i += 1 + word[0].length
        continue
      }
    }

    buffer += src[i]
    i++
  }

  flush()
  return out
}

/** Distinct mention labels in `text`, in order of first appearance. */
export function extractMentionLabels(
  text: string,
  matcher: MentionMatcher | null = null,
): string[] {
  const seen = new Set<string>()
  const out: string[] = []
  for (const seg of parseMentionSegments(text, matcher)) {
    if (seg.type !== 'mention') continue
    const key = seg.label.toLowerCase()
    if (seen.has(key)) continue
    seen.add(key)
    out.push(seg.label)
  }
  return out
}

/**
 * Rewrite every mention in `text` through `render`, leaving literal runs alone.
 * Used to turn stored markdown into chip markup.
 */
export function replaceMentions(
  text: string,
  matcher: MentionMatcher | null,
  render: (label: string) => string,
): string {
  return parseMentionSegments(text, matcher)
    .map(seg => (seg.type === 'mention' ? render(seg.label) : seg.value))
    .join('')
}

/** True when `name` must be written as `@"name"` to survive a bare parse. */
export function mentionNeedsQuotes(name: string | null | undefined): boolean {
  return !!name && NEEDS_QUOTES_RE.test(name)
}

/** Canonical source form for a mention of `name`. */
export function formatMention(name: string): string {
  return mentionNeedsQuotes(name) ? `@"${name.replace(/"/g, '')}"` : `@${name}`
}

// ---------------------------------------------------------------------------
// Prompt mentions -> InstructionText references
// ---------------------------------------------------------------------------

/** Prompt mention groups are keyed by display label; references by object type. */
const GROUP_TYPE_MAP: Record<string, string> = {
  'DATA SOURCES': 'data_source',
  'TABLES': 'datasource_table',
  'FILES': 'file',
  'ENTITIES': 'entity',
  'CONNECTION TOOLS': 'connection_tool',
}

export interface PromptMentionRef {
  id: string
  type: string
  name: string
  data_source_type?: string
}

/**
 * Flatten a completion prompt's `mentions` groups into the reference list
 * InstructionText matches against, so an @mention in a user prompt renders as a
 * chip rather than bare text. Shared by the report view and the public share
 * view — the two render the same prompts and must resolve them identically.
 */
export function promptMentionsToRefs(
  mentions?: Array<{ name: string; items: any[] }>,
): PromptMentionRef[] {
  if (!mentions?.length) return []
  const refs: PromptMentionRef[] = []
  for (const group of mentions) {
    const type = GROUP_TYPE_MAP[(group.name || '').toUpperCase()] || 'entity'
    for (const item of group.items || []) {
      let name = item.name || item.title || item.filename || ''
      // Data-source tables are serialized into the prompt text with their
      // source prefix (e.g. "@Microsoft Fabric / dbo.sales"), so the ref
      // name must include it to match and render as a single mention chip.
      if (type === 'datasource_table') {
        const prefix = item.connection_name || item.data_source_name
        if (prefix) name = `${prefix} / ${name}`
      }
      refs.push({
        id: item.id,
        type,
        // `icon_type` is what the mention items actually carry (set to the
        // data source's type); include it so the chip renders the correct
        // data-source icon instead of the generic fallback glyph.
        data_source_type: item.connection_type || item.data_source_type || item.icon_type || undefined,
        name,
      })
    }
  }
  return refs
}
